### PR TITLE
Adding first saving/reload tests

### PR DESCRIPTION
**Pull request by rbharath**
In this PR, I'm working on adding saving/reload tests for our models. It turns out that a number of models fail to reload correctly. I haven't yet added tests for all our models, but so far here are the failing models:

- `WeaveModel`
- `GraphConvModel`
- `TextCNNModel`
- `MPNNModel`
- `DAGModel`

I thought that only `GraphConvModel` had saving issues but it looks like we have additional issues. I've updated our coding guidelines as part of this PR to ask that all new models have overfit and reloading unit tests upon addition. I'll continue iterating on this PR until I figure out what's causing the reloading issues

### CODE REVIEW

**Comment by rbharath** [comment]
CC @peastman @nd-02110114 @ncfrey @seyonechithrananda

**Comment by rbharath** [comment]
I'm looking at the save/reload behavior for `WeaveModel` and I'm seeing some interesting behavior. For the saved/reloaded model, all weights for the underlying model match. These asserts pass:

```
for i in range(len(model.model.get_weights())): 
  assert np.all(model.model.get_weights()[i] == reloaded_model.model.get_weights()[i]) 
```
However the actual prediction test fails.
```
np.all(model.predict(predset) == reloaded_model.predict(predset)) 
```

This is a little interesting since the weightss match but the predictions don't.

**Comment by rbharath** [comment]
@peastman I'm trying to understand what could be causing these failures. So far, the common pattern I see is that the models with our custom Keras layers are failing to reload properly. Is there a good way to test serialization/deserialization of a keras layer? I think the KerasModel restoration code itself is correct (it's very simple), which suggests the failure should exist in the custom keras layers

**Comment by rbharath** [comment]
Here's an interesting code snippet. Here I save/restore a model from disk using the keras save/load rather than the `KerasModel.restore()` and find that restoration works correctly

```
  featurizer = dc.feat.WeaveFeaturizer()
  tasks = ["outcome"]
  mols = ["C", "CO", "CC"]
  n_samples = len(mols)
  X = featurizer(mols)
  y = np.random.randint(2, size=(n_samples, n_tasks))
  dataset = dc.data.NumpyDataset(X, y)

  classification_metric = dc.metrics.Metric(dc.metrics.roc_auc_score)

  batch_size = 10

  model_dir = tempfile.mkdtemp()
  model = dc.models.WeaveModel(
      n_tasks,
      batch_size=batch_size,
      learning_rate=0.0003,
      mode="classification",
      dropouts=0.0,
      model_dir=model_dir)

  # Fit trained model
  model.fit(dataset, nb_epoch=30)

  # Eval model on train
  scores = model.evaluate(dataset, [classification_metric])
  assert scores[classification_metric.name] > .9

  # Custom save
  save_dir = tempfile.mkdtemp()
  model.model.save(save_dir)

  from tensorflow import keras
  reloaded = keras.models.load_model(save_dir)

  reloaded_model = dc.models.WeaveModel(
      n_tasks,
      batch_size=batch_size,
      learning_rate=0.0003,
      mode="classification",
      dropouts=0.0,
      model_dir=model_dir)

  # Don't restore as usual! Instead directly place in reloaded keras model
  #reloaded_model.restore()
  reloaded_model.model = reloaded

  # Check predictions match on random sample
  predmols = ["CCCC", "CCCCCO", "CCCCC"]
  Xpred = featurizer(predmols)
  predset = dc.data.NumpyDataset(Xpred)
  origpred = model.predict(predset)
  reloadpred = reloaded_model.predict(predset)
  # This test passes
  assert np.all(origpred == reloadpred)
```

This code snippet passes. However, if I use `model.restore()` it fails! I wonder if somehow the checkpoint manager isn't updating the underlying model correctly. Still debugging here

**Comment by rbharath** [comment]
Continuing to debug, I find some interesting behavior that it's clear something is being reloaded for most of these models. For `WeaveModel` and `MPNNModel`, the reloaded model often has accuracy similar to the original model on the small dataset the original model was trained on. However, predictions are not identical and in some cases can differ quite a bit.

My working hypothesis is that when `WeaveModel` or `MPNNModel` etc are instantiated, there are parts of the instantiation that are not captured in the checkpoint somehow. I'm still pretty confused but will keep grinding on this one

**Comment by rbharath** [comment]
Fixing the serialization tests is going to take some work. I think it would be better to get a first round of correctness tests in on the simpler models and then debug the more complex models one by one. I'll put up a new issue to track these fixes. In the meanwhile, once tests pass, I think we should go ahead and merge this in. (There's some commented out code in the tests that I'll clean up in subsequent PRs as I continue work on serialization this week.)

**Comment by rbharath** [comment]
Ok this PR is green, so going ahead and merging in. Will move on to hitting more of the models listed in #2217.
